cpp/src/parquet/variant.h

```diff
@@ -0,0 +1,261 @@
+
+enum class VariantPrimitiveType : int8_t {
+  /// Equivalent Parquet Type: UNKNOWN
+  NullType = 0,
+  /// Equivalent Parquet Type: BOOLEAN
+  BooleanTrue = 1,
+  /// Equivalent Parquet Type: BOOLEAN
+  BooleanFalse = 2,
+  /// Equivalent Parquet Type: INT(8, signed)
+  Int8 = 3,
+  /// Equivalent Parquet Type: INT(16, signed)
+  Int16 = 4,
+  /// Equivalent Parquet Type: INT(32, signed)
+  Int32 = 5,
+  /// Equivalent Parquet Type: INT(64, signed)
+  Int64 = 6,
+  /// Equivalent Parquet Type: DOUBLE
+  Double = 7,
+  /// Equivalent Parquet Type: DECIMAL(precision, scale)
+  Decimal4 = 8,
+  /// Equivalent Parquet Type: DECIMAL(precision, scale)
+  Decimal8 = 9,
+  /// Equivalent Parquet Type: DECIMAL(precision, scale)
+  Decimal16 = 10,
+  /// Equivalent Parquet Type: DATE
+  Date = 11,
+  /// Equivalent Parquet Type: TIMESTAMP(isAdjustedToUTC=true, MICROS)
+  Timestamp = 12,
+  /// Equivalent Parquet Type: TIMESTAMP(isAdjustedToUTC=false, MICROS)
+  TimestampNTZ = 13,
+  /// Equivalent Parquet Type: FLOAT
+  Float = 14,
+  /// Equivalent Parquet Type: BINARY
+  Binary = 15,
+  /// Equivalent Parquet Type: STRING
+  String = 16,
+  /// Equivalent Parquet Type: TIME(isAdjustedToUTC=false, MICROS)
+  TimeNTZ = 17,
+  /// Equivalent Parquet Type: TIMESTAMP(isAdjustedToUTC=true, NANOS)
+  TimestampTZ = 18,  // Assuming TZ stands for TimeZone, and follows the document's
+                     // 'timestamp with time zone'
+  /// Equivalent Parquet Type: TIMESTAMP(isAdjustedToUTC=false, NANOS)
+  TimestampNTZNanos = 19,  // Differentiating from TimestampNTZ (MICROS)
+  /// Equivalent Parquet Type: UUID
+  Uuid = 20
+};
+
+std::string variantPrimitiveTypeToString(VariantPrimitiveType type);
+
+/// VariantType is from basic type and primitive type.
+enum class VariantType {
+  OBJECT,
+  ARRAY,
+  VARIANT_NULL,
+  BOOLEAN,
+  INT8,
+  INT16,
+  INT32,
+  INT64,
+  STRING,
+  DOUBLE,
+  DECIMAL4,
+  DECIMAL8,
+  DECIMAL16,
+  DATE,
+  TIMESTAMP_TZ,
+  TIMESTAMP_NTZ,
+  FLOAT,
+  BINARY,
+  TIME,
+  TIMESTAMP_NANOS_TZ,
+  TIMESTAMP_NANOS_NTZ,
+  UUID
+};
+
+std::string variantTypeToString(VariantType type);
+
+class VariantMetadata {
+ public:
+  explicit VariantMetadata(std::string_view metadata);
+  /// \brief Get the variant metadata version. Currently, always 1.
+  uint8_t version() const;
+  /// \brief Get the metadata key for a given variant field id.
+  /// \throw ParquetException if the variant_id is out of range(larger than
+  ///        dictionary size).
+  std::string_view getMetadataKey(uint32_t variant_id) const;
+  /// \brief Get the metadata id for a given key.
+  /// From the discussion in ML:
+  /// https://lists.apache.org/thread/b68tjmrjmy64mbv9dknpmqs28vnzjj96 if
+  /// !sortedStrings(), the metadata key is not guaranteed to be unique, so we use a
+  /// vector to store all the metadata ids.
+  ::arrow::internal::SmallVector<uint32_t, 1> getMetadataId(std::string_view key) const;
+
+  bool sortedStrings() const;
```

Comment:
```suggestion
  bool sorted_strings() const;
```

Or `strings_sorted` looks more nature to me...